mailpoet/lib/Automation/Engine/Control/StepRunLogger.php

```diff
@@ -0,0 +1,85 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Engine\Control;
+
+use DateTimeImmutable;
+use MailPoet\Automation\Engine\Data\AutomationRunLog;
+use MailPoet\Automation\Engine\Hooks;
+use MailPoet\Automation\Engine\Storage\AutomationRunLogStorage;
+use MailPoet\InvalidStateException;
+use Throwable;
+
+class StepRunLogger {
+  /** @var AutomationRunLogStorage */
+  private $automationRunLogStorage;
+
+  /** @var Hooks */
+  private $hooks;
+
+  /** @var int */
+  private $runId;
+
+  /** @var string */
+  private $stepId;
+
+  /** @var AutomationRunLog|null */
+  private $log;
+
+  public function __construct(
+    AutomationRunLogStorage $automationRunLogStorage,
+    Hooks $hooks,
+    int $runId,
+    string $stepId
+  ) {
+    $this->automationRunLogStorage = $automationRunLogStorage;
+    $this->hooks = $hooks;
+    $this->runId = $runId;
+    $this->stepId = $stepId;
+  }
+
+  public function logStart(): void {
+    $this->getLog();
+  }
+
+  public function logSuccess(): void {
+    $log = $this->getLog();
+    $log->setStatus(AutomationRunLog::STATUS_COMPLETE);
+    $log->setCompletedAt(new DateTimeImmutable());
+    $this->triggerAfterRunHook($log);
+    $this->automationRunLogStorage->updateAutomationRunLog($log);
+  }
+
+  public function logFailure(Throwable $error): void {
+    $log = $this->getLog();
+    $log->setStatus(AutomationRunLog::STATUS_FAILED);
+    $log->setError($error);
+    $log->setCompletedAt(new DateTimeImmutable());
+    $this->triggerAfterRunHook($log);
+    $this->automationRunLogStorage->updateAutomationRunLog($log);
+  }
+
+  private function getLog(): AutomationRunLog {
+    if (!$this->log) {
+      $this->log = $this->automationRunLogStorage->getAutomationRunLogByRunAndStepId($this->runId, $this->stepId);
+    }
+
+    if (!$this->log) {
+      $log = new AutomationRunLog($this->runId, $this->stepId);
+      $id = $this->automationRunLogStorage->createAutomationRunLog($log);
+      $this->log = $this->automationRunLogStorage->getAutomationRunLog($id);
+    }
+
+    if (!$this->log) {
+      throw new InvalidStateException('Failed to create automation run log');
+    }
+    return $this->log;
+  }
+
+  private function triggerAfterRunHook(AutomationRunLog $log): void {
+    try {
+      $this->hooks->doAutomationStepAfterRun($log);
+    } catch (Throwable $e) {
+      // ignore integration errors
```

Comment:
Hmmm.... Why do we `ignore integration errors`?

I see we want to somewhat protect our runs and ensure we finish each step and log properly etc.pp.

But I am just wondering, if this might be risky too when it comes to debugging :thinking: 

So, lets say I build some 

```
add_action('<step_run_after>', function(){
 //oups
 throw Error();
});
```

this will be silenced and i might overlook it during development. I mean, the only place I would see those errors would be in php logs, `debug.log` or something (as its async), but there i would see them otherwise usually.

Maybe we could do something like

`try {
// some integration
} catch( Throwable $e) {
 if (defined(WP_DEBUG)) {
 throw $e;
 }
}
```

which makes me think (I think i saw this "ignore integration errors" somewhere else) if we should handle this generally in `Hooks` instead. Even if we conclude in the end to just ignore, it looks to me like this is something which would be better handled there :thinking:

Reply:
@websupporter I kept this as it was in the original code, only moved it up. This is only a case when an integration tries to add some logging data and fails to do so, although the hook should probably be renamed to reflect that. I can also add the `WP_DEBUG` option, makes sense 👍